includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -978,22 +971,23 @@ public function process_payment_for_order( $cart, $payment_information, $additio
 			}
 
 			if ( empty( $intent ) ) {
-				// Create intention, try to confirm it & capture the charge (if 3DS is not required).
-				$intent = $this->payments_api_client->create_and_confirm_intention(
-					$converted_amount,
-					$currency,
-					$payment_information->get_payment_method(),
-					$customer_id,
-					$payment_information->is_using_manual_capture(),
-					$save_payment_method_to_store,
-					$payment_information->should_save_payment_method_to_platform(),
-					$metadata,
-					$this->get_level3_data_from_order( $order ),
-					$payment_information->is_merchant_initiated(),
-					$additional_api_parameters,
-					$payment_methods,
-					$payment_information->get_cvc_confirmation()
-				);
+				add_filter( 'woopay_create_and_confirm_intention_request_from_base_intent_request', [ Checkout_Service::class, 'create_woopay_intention_request' ], 10, 3 );
+				$request = WC_Payments::create_request( Create_And_Confirm_Intention::class )
+					->set_amount( $converted_amount )
+					->set_currency_code( $currency )
+					->set_payment_method( $payment_information->get_payment_method() )
+					->set_customer( $customer_id )
+					->set_capture_method( $payment_information->is_using_manual_capture() )
+					->set_metadata( $metadata )
+					->set_level3( $this->get_level3_data_from_order( $order ) )
+					->set_off_session( $payment_information->is_merchant_initiated() )
+					->set_payment_methods( $payment_methods )
+					->set_cvc_confirmation( $payment_information->get_cvc_confirmation() );
+
+				if ( $save_payment_method_to_store ) {
+					$request->setup_future_usage();
+				}
+				$intent = $request->send( 'create_woopay_intention_request', $order, $this->is_platform_payment_method( $payment_information->is_using_saved_payment_method() ) );
```

Comment:
`$this->is_platform_method()` is a WooPay method, and should not remain in the gateway. As a follow-up, and partially fixing https://github.com/automattic/woocommerce-payments/issues/4953, we should move out all WooPay-related method from this class, to the new service. This includes not only `is_platform_payment_method`, but also ones like `should_use_stripe_platform_on_checkout_page()`, and whatever they are modifying.

I'd like the one example we have of `$request->send()` and its filter to be a proper one, which would not need to change in a version or two of the gateway, when `should_use_stripe_platform_on_checkout_page`-like methods get extracted from this class.

I'm OK with making `is_platform_payment_method` public for now, and using it within `WCPay\WooPay\Service\Checkout_Service` instead of moving it, but this line should not contain anything woopay-related.

Actually, why did you change the filter to `create_woopay_intention_request`?! This is not a WooPay create intention request, it's a **WCPay** create intention request?

- The idea is to have WooPay use the same set of functionality, available for all other extensions.
- Other developers might think that this is a WooPay-specific filter and restrain from using it, even though it is not.
- By the way, this is supposed to be `wcpay_create_and_confirm_intention_request`, what happened to the `_and_confirm` part? There are other requests for `create_intention`, and it's still something that is and will be used.

Reply:
This is resolved with new commits. The filter naming is returned and some of the functions are for now extracted to specific WooPay service. To maintain backward compatibility, there could be a duplicate code, but we can easily move it once we migrate all API requests to request classes.

The filter part is also extracted to a different (custom) WooPay service with brand new folder and namespace.